central/networkgraph/service/service_impl.go

```diff
@@ -101,24 +100,60 @@ func (s *serviceImpl) GetExternalNetworkEntities(ctx context.Context, request *v
 	}
 
 	query, _ = search.FilterQueryWithMap(query, schema.NetworkEntitiesSchema.OptionsMap)
-	pred, err := netEntityPredFactory.GeneratePredicate(query)
+
+	ets, err := s.entities.GetEntityByQuery(ctx, query)
 	if err != nil {
-		return nil, errors.Wrapf(errox.InvalidArgs, "failed to parse query %q: %v", query.String(), err.Error())
+		return nil, err
 	}
 
-	ret, err := s.entities.GetAllMatchingEntities(ctx, func(entity *storage.NetworkEntity) bool {
-		// Do not respect the graph configuration.
-		if entity.GetScope().GetClusterId() == "" || entity.GetScope().GetClusterId() == request.GetClusterId() {
-			return pred.Matches(entity)
-		}
-		return false
-	})
+	return &v1.GetExternalNetworkEntitiesResponse{
+		Entities: ets,
+	}, nil
+}
+
+func (s *serviceImpl) GetExternalNetworkFlows(ctx context.Context, request *v1.GetExternalNetworkFlowsRequest) (*v1.GetExternalNetworkFlowsResponse, error) {
+	// Temporarily elevate permissions to obtain all network flows in cluster.
+	networkGraphGenElevatedCtx := sac.WithGlobalAccessScopeChecker(context.Background(),
+		sac.AllowFixedScopes(
+			sac.AccessModeScopeKeys(storage.Access_READ_ACCESS),
+			sac.ResourceScopeKeys(resources.NetworkGraph),
+			sac.ClusterScopeKeys(request.GetClusterId())))
+
+	flowStore, err := s.getFlowStore(networkGraphGenElevatedCtx, request.GetClusterId())
 	if err != nil {
 		return nil, err
 	}
 
-	return &v1.GetExternalNetworkEntitiesResponse{
-		Entities: ret,
+	flows, err := flowStore.GetExternalFlowsForDeployment(ctx, request.GetDeploymentId())
+	if err != nil {
+		return nil, err
+	}
+
+	// Populate entities. This could probably be implemented in the DB query
+	// for better efficiency, but is simply implemented here for now.
+	for _, flow := range flows {
+		var toGet string
+		var toSet **storage.NetworkEntityInfo
+
+		if flow.GetProps().GetSrcEntity().GetType() == storage.NetworkEntityInfo_EXTERNAL_SOURCE {
```

Comment:
Save `flow.GetProps().GetSrcEntity()` and `flow.GetProps().GetDstEntity()` to variables. Also you probably need to check for nil.

Reply:
+1 to the nil check, looks like `GetProps()`, `GetSrcEntity()` and `GetDstEntity()` can all return nil, so we probably want to check it unless we have a strong reason why we shouldn't.